diff --git a/vagrantfiles/Vagrantfile b/vagrantfiles/Vagrantfile
index 834eda5..83522c0 100644
--- a/vagrantfiles/Vagrantfile
+++ b/vagrantfiles/Vagrantfile
@@ -42,7 +42,7 @@ load $osvariablefile
 Vagrant.configure('2') do |config|
     config.vm.box = $box_image
     config.vm.box_check_update = true
-
+    config.disksize.size = '20GB'
     config.vm.boot_timeout = 900
 
     config.vm.provider VAGRANT_DEFAULT_PROVIDER do |l|
